keel-web/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -0,0 +1,169 @@
+      environmentSummary.artifacts.map { artifact ->
+        artifact.key to mutableSetOf<ArtifactVersionSummary>()
+      }
+    }.toMap()
+
+    // for each environment...
+    val artifactSummaries = environmentSummaries.flatMap { environmentSummary ->
+
+      // ...and each artifact in the environment...
+      environmentSummary.artifacts.map { artifact ->
+        val versionsByState = mapOf(
+          "current" to artifact.versions.current?.let { listOf(it) }.orEmpty(),
+          "deploying" to artifact.versions.deploying?.let { listOf(it) }.orEmpty(),
+          "pending" to artifact.versions.pending,
+          "approved" to artifact.versions.approved,
+          "previous" to artifact.versions.previous,
+          "vetoed" to artifact.versions.vetoed
+        )
+
+        // ...loop the artifact versions by state...
+        versionsByState.entries.map {
+          val state = it.key
+          val versions = it.value
+
+          // ...then record the artifact version summary for each version for that state...
+          if (versions != null) {
+            versionSummariesByArtifact[artifact.key]!!.addAll(
+              // ...for each version...
+              versions.map { version ->
+                // ...get an artifact summary for this version in this environment...
+                val summaryInEnvironment = artifactRepository.getArtifactSummaryInEnvironment(
+                  deliveryConfig = deliveryConfig,
+                  environmentName = environmentSummary.name,
+                  artifactName = artifact.name,
+                  artifactType = artifact.type,
+                  version = version
+                )
+
+                // ...create the artifact version summary for this version and add to the list
+                ArtifactVersionSummary(
+                  version = version,
+                  environments = artifactSummariesByEnvironmentAndState
+                    .get(environmentSummary.name)!! // safe because we create the maps with these keys above
+                    .get(state)!! // safe because we create the maps with these keys above
+                    .also { artifactSummariesInEnvironment ->
+                      if (summaryInEnvironment != null) {
+                        artifactSummariesInEnvironment.add(summaryInEnvironment)
+                      }
+                    }
+                )
+              }
+            )
+          }
+        }
+        // finally, create the artifact summary by looking up the version summaries that were built above
+        ArtifactSummary(
+          name = artifact.name,
+          type = artifact.type,
+          versions = versionSummariesByArtifact[artifact.key]
+        )
+      }
+    }
+
+    return artifactSummaries
+  }
+
+  private fun getFirstDeliveryConfigFor(application: String): DeliveryConfig? =
+    repository.getDeliveryConfigsByApplication(application).also {
```

Comment:
Again, upon discussion, we agreed to assume a single delivery config per application for now.